File: tests/pooltowin.ts
import * as anchor from "@project-serum/anchor";
import { Program } from "@project-serum/anchor";
import { createAssociatedTokenAccount, getAccount } from "@solana/spl-token";
import { expect } from "chai";
import { Pooltowin } from "../target/types/pooltowin";

describe("pooltowin", () => {
	// Configure the client to use the local cluster.
	const provider = anchor.AnchorProvider.env();
	anchor.setProvider(provider);

	const program = anchor.workspace.Pooltowin as Program<Pooltowin>;

	const owner = provider.wallet.publicKey;

	const user = anchor.web3.Keypair.generate();

	const getPDAFromSeed = (seed: string) => {
		return anchor.web3.PublicKey.findProgramAddressSync(
			[Buffer.from(seed), owner.toBuffer()],
			program.programId
		)[0];
	};

	const poolPDA = getPDAFromSeed("pool-account");
	const tokenAccountPDA = getPDAFromSeed("token-account");
	const tokenMintPDA = getPDAFromSeed("token-mint");
	const userInfoPDA = getPDAFromSeed("user-info");

	let userTokenAccount: anchor.web3.PublicKey;

	it("Setup", async () => {
		await provider.connection.requestAirdrop(
			user.publicKey,
			anchor.web3.LAMPORTS_PER_SOL
		);
	});

	it("Creates new token", async () => {
		await program.methods
			.initTokenMint()
			.accounts({ owner, tokenMint: tokenMintPDA })
			.rpc();
		console.log(`New token created`);
	});

	it("Creates pool", async () => {
		await program.methods
			.initPool()
			.accounts({
				owner,
				pool: poolPDA,
				tokenAccount: tokenAccountPDA,
				mint: tokenMintPDA,
			})
			.rpc();
		console.log(`New pool created`);
	});

	it("Creates user token", async () => {
		userTokenAccount = await createAssociatedTokenAccount(
			provider.connection,
			user,
			tokenMintPDA,
			user.publicKey
		);
		console.log(`User token account: ${userTokenAccount}`);
	});

	it("Mints token", async () => {
		await program.methods
			.mintToken()
			.accounts({
				mint: tokenMintPDA,
				recipient: userTokenAccount,
				signer: owner,
			})
			.rpc();
		const userAccount = await getAccount(provider.connection, userTokenAccount);
		console.log(`User token amount: ${userAccount.amount}`);

		await program.methods
			.mintToken()
			.accounts({
				mint: tokenMintPDA,
				recipient: tokenAccountPDA,
				signer: owner,
			})
			.rpc();
		const poolAccount = await getAccount(provider.connection, tokenAccountPDA);
		console.log(`Pool token amount: ${poolAccount.amount}`);
	});
	
	it("Creates user info", async () => {
		await program.methods.initUserInfo().accounts({
			owner,
			userInfo: userInfoPDA,
		}).rpc();

		const userInfo = await program.account.userInfo.fetch(userInfoPDA);
		expect(userInfo.amount.toString()).to.equal("0");
	})

	it("Deposits token in pool", async () => {
		const depositAmount = new anchor.BN(100);
		await program.methods.depositPool(depositAmount).accounts({
			signer: user.publicKey,
			userTokenAccount,
			poolTokenAccount: tokenAccountPDA,
			userInfo: userInfoPDA
		}).signers([user]).rpc();

		const userAccount = await getAccount(provider.connection, userTokenAccount);
		expect(userAccount.amount.toString()).to.equal("900");
		const poolAccount = await getAccount(provider.connection, tokenAccountPDA);
		expect(poolAccount.amount.toString()).to.equal("1100");
		const userInfoAccount = await program.account.userInfo.fetch(userInfoPDA);
		expect(userInfoAccount.amount.toString()).to.equal("100");
	});

	it("Withdraws token from pool", async () => {
		const withdrawAmount = new anchor.BN(100);
		await program.methods.withdrawPool(withdrawAmount).accounts({
			signer: owner,
			userTokenAccount,
			poolTokenAccount: tokenAccountPDA,
			userInfo: userInfoPDA
		}).rpc();

		const userAccount = await getAccount(provider.connection, userTokenAccount);
		expect(userAccount.amount.toString()).to.equal("1000");
		const poolAccount = await getAccount(provider.connection, tokenAccountPDA);
		expect(poolAccount.amount.toString()).to.equal("1000");
		const userInfoAccount = await program.account.userInfo.fetch(userInfoPDA);
		expect(userInfoAccount.amount.toString()).to.equal("0");
	});
});

File: programs_py/pooltowin.py
# pooltowin
# Built with Seahorse v0.1.1

from curses import use_default_colors
from tokenize import Token

from seahorse.prelude import *

declare_id('FPpwkb2FsmhYnUXfbrxkE5HpwKZsbCcwExDUudrVdGdY')


class Pool(Account):
    token_mint: Pubkey
    token_account: Pubkey
    owner: Pubkey

class UserInfo(Account):
    amount: u64

@instruction
def init_token_mint(owner: Signer, token_mint: Empty[TokenMint]):
    token_mint.init(
        payer=owner,
        seeds=["token-mint", owner],
        decimals=6,
        authority=owner
    )

@instruction
def init_user_info(owner: Signer, user_info: Empty[UserInfo]):
    user_info.init(
        payer=owner,
        seeds=["user-info", owner],
    )
    user_info.amount = 0

@instruction
def init_pool(owner: Signer, pool: Empty[Pool], token_account: Empty[TokenAccount], mint: TokenMint):
    pool.init(
        payer=owner,
        seeds=["pool-account", owner],
    )
    pool.owner = owner.key()
    pool.token_mint = mint.key()
    pool.token_account = token_account.init(
        payer=owner,
        seeds=["token-account", owner],
        mint=mint,
        authority=owner
    ).key()


@instruction
def mint_token(mint: TokenMint, recipient: TokenAccount, signer: Signer):
    mint.mint(
        authority=signer,
        to=recipient,
        amount=1000
    )


@instruction
def deposit_pool(signer: Signer, user_token_account: TokenAccount, pool_token_account: TokenAccount, user_info: UserInfo, amount: u64):
    assert amount > 0, "Deposit amount should be more than 0"

    user_token_account.transfer(
        authority=signer,
        to=pool_token_account,
        amount=amount
    )

    user_info.amount += amount


@instruction
def withdraw_pool(signer: Signer, user_token_account: TokenAccount, pool_token_account: TokenAccount, user_info: UserInfo, amount: u64):
    assert amount > 0, "Withdraw amount should be more than 0"

    pool_token_account.transfer(
        authority=signer,
        to=user_token_account,
        amount=amount
    )

    user_info.amount -= amount
